File: dashboard/backend/evaluationplane/method_contract_v2.go
package evaluationplane

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"reflect"
	"sort"
	"strings"
)

// EvaluationMethodContractVersion is intentionally independent of the v1 run
// bundle contract.  No v1 method record is silently promoted into this model.
const (
	EvaluationMethodContractVersion = "evaluation-method.v2"
	R2CompoundModelBudgetMethodID   = "r2.compound-model-budget.v2"
)

type ActionRef struct {
	SchemaVersion string `json:"schema_version"`
	ID            string `json:"id"`
}

type SliceRef struct {
	SchemaVersion string `json:"schema_version"`
	ID            string `json:"id"`
}

type AnalysisPlan struct {
	SchemaVersion string     `json:"schema_version"`
	ID            string     `json:"id"`
	AnalysisUnit  string     `json:"analysis_unit"`
	ClusterUnit   string     `json:"cluster_unit"`
	Slices        []SliceRef `json:"slices"`
	CurveDomain   string     `json:"curve_domain"`
	Missingness   string     `json:"missingness"`
}

// EvaluationMethodDefinition makes the execution and evidentiary limits of an
// advertised method explicit.  A live method is only cataloguable when both
// its input and its grading are complete.
type EvaluationMethodDefinition struct {
	SchemaVersion       string        `json:"schema_version"`
	ID                  string        `json:"id"`
	Version             string        `json:"version"`
	Status              string        `json:"status"`
	ExecutionOwner      string        `json:"execution_owner"`
	InputSchema         string        `json:"input_schema"`
	ExportSchema        string        `json:"export_schema"`
	LiveInputComplete   bool          `json:"live_input_complete"`
	LiveGrader          bool          `json:"live_grader"`
	ApplicableTracks    []TrackID     `json:"applicable_tracks"`
	LiveTracks          []TrackID     `json:"live_tracks"`
	ProducedMetricIDs   []string      `json:"produced_metric_ids"`
	EvidenceCeiling     EvidenceLevel `json:"evidence_ceiling"`
	NativeParity        string        `json:"native_parity"`
	RequiredArtifactIDs []string      `json:"required_artifact_ids"`
	AnalysisPlan        AnalysisPlan  `json:"analysis_plan"`
}

var evaluationMethodRequiredFields = [...]string{
	"schema_version",
	"id",
	"version",
	"status",
	"execution_owner",
	"input_schema",
	"export_schema",
	"live_input_complete",
	"live_grader",
	"applicable_tracks",
	"live_tracks",
	"produced_metric_ids",
	"evidence_ceiling",
	"native_parity",
	"required_artifact_ids",
	"analysis_plan",
}

// UnmarshalJSON is the production wire boundary for method declarations. Go
// value fields otherwise collapse an omitted or null boolean/list into the same
// zero value as an explicit false/empty declaration. Preserve typed registry
// construction while making every JSON declaration prove the complete v2
// contract before it can enter a report or admission path.
func (method *EvaluationMethodDefinition) UnmarshalJSON(data []byte) error {
	if err := rejectDuplicateJSONKeys(data); err != nil {
		return fmt.Errorf("decode evaluation method: %w", err)
	}
	var fields map[string]json.RawMessage
	fieldDecoder := json.NewDecoder(bytes.NewReader(data))
	if err := fieldDecoder.Decode(&fields); err != nil {
		return fmt.Errorf("decode evaluation method fields: %w", err)
	}
	if err := ensureJSONEOF(fieldDecoder); err != nil {
		return err
	}
	for _, name := range evaluationMethodRequiredFields {
		value, present := fields[name]
		if !present {
			return fmt.Errorf("evaluation method field %q is required", name)
		}
		if bytes.Equal(bytes.TrimSpace(value), []byte("null")) {
			return fmt.Errorf("evaluation method field %q cannot be null", name)
		}
	}

	type methodDefinitionWire EvaluationMethodDefinition
	var decoded methodDefinitionWire
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&decoded); err != nil {
		return fmt.Errorf("decode evaluation method: %w", err)
	}
	if err := ensureJSONEOF(decoder); err != nil {
		return err
	}
	candidate := EvaluationMethodDefinition(decoded)
	if err := ValidateEvaluationMethodDefinition(candidate); err != nil {
		return err
	}
	*method = candidate
	return nil
}

func validateActionRef(ref ActionRef) error {
	if ref.SchemaVersion != EvaluationMethodContractVersion || !validMethodID(ref.ID) {
		return fmt.Errorf("action ref is invalid")
	}
	return nil
}

func validateSliceRef(ref SliceRef) error {
	if ref.SchemaVersion != EvaluationMethodContractVersion || !validMethodID(ref.ID) {
		return fmt.Errorf("slice ref is invalid")
	}
	return nil
}

func validateAnalysisPlan(plan AnalysisPlan) error {
	if plan.SchemaVersion != EvaluationMethodContractVersion || !validMethodID(plan.ID) ||
		strings.TrimSpace(plan.AnalysisUnit) == "" || strings.TrimSpace(plan.ClusterUnit) == "" ||
		(plan.CurveDomain != "shared_budget" && plan.CurveDomain != "not_applicable") ||
		plan.Missingness != "fail_closed" || len(plan.Slices) == 0 {
		return fmt.Errorf("analysis plan is invalid")
	}
	seen := make(map[string]struct{}, len(plan.Slices))
	for _, slice := range plan.Slices {
		if err := validateSliceRef(slice); err != nil {
			return err
		}
		if _, duplicate := seen[slice.ID]; duplicate {
			return fmt.Errorf("analysis plan slices must be unique")
		}
		seen[slice.ID] = struct{}{}
	}
	return nil
}

func validEvidenceCeiling(value EvidenceLevel) bool {
	for _, ceiling := range []EvidenceLevel{"E0", "E1", "E2", "E3", "E4", "E5"} {
		if value == ceiling {
			return true
		}
	}
	return false
}

func validMethodTrackID(value TrackID) bool {
	for _, track := range allTrackIDs {
		if value == track {
			return true
		}
	}
	return false
}

// ValidateEvaluationMethodDefinition is the catalog/planner admission boundary.
func ValidateEvaluationMethodDefinition(method EvaluationMethodDefinition) error {
	owners := map[string]bool{"server": true, "worker": true, "provider": true, "benchmark_native": true}
	parities := map[string]bool{"native": true, "source_qualified": true, "none": true}
	statuses := map[string]bool{"native-qualified": true, "exploratory-import": true, "data-required": true, "blocked": true}
	if method.SchemaVersion != EvaluationMethodContractVersion || method.Version != EvaluationMethodContractVersion ||
		!validMethodID(method.ID) || !statuses[method.Status] || !owners[method.ExecutionOwner] || !validMethodID(method.InputSchema) ||
		!validMethodID(method.ExportSchema) ||
		!validEvidenceCeiling(method.EvidenceCeiling) || !parities[method.NativeParity] {
		return fmt.Errorf("evaluation method definition is invalid")
	}
	if method.NativeParity == "native" && method.ExecutionOwner != "benchmark_native" {
		return fmt.Errorf("native method parity requires benchmark-native execution")
	}
	if len(method.ApplicableTracks) == 0 || len(method.ProducedMetricIDs) == 0 {
		return fmt.Errorf("evaluation method definition needs produced metrics")
	}
	if len(method.RequiredArtifactIDs) == 0 {
		return fmt.Errorf("evaluation method definition needs required artifacts")
	}
	if method.Status == "native-qualified" && (!method.LiveInputComplete || !method.LiveGrader || len(method.LiveTracks) == 0) {
		return fmt.Errorf("native-qualified method needs complete graded live execution")
	}
	if method.Status != "native-qualified" && (method.LiveInputComplete || method.LiveGrader) {
		return fmt.Errorf("non-qualified method cannot claim complete live execution")
	}
	seenApplicableTracks := make(map[TrackID]struct{}, len(method.ApplicableTracks))
	for _, track := range method.ApplicableTracks {
		if !validMethodTrackID(track) {
			return fmt.Errorf("evaluation method applicable track is invalid")
		}
		if _, duplicate := seenApplicableTracks[track]; duplicate {
			return fmt.Errorf("evaluation method applicable tracks must be unique")
		}
		seenApplicableTracks[track] = struct{}{}
	}
	seenTracks := make(map[TrackID]struct{}, len(method.LiveTracks))
	for _, track := range method.LiveTracks {
		if !validMethodTrackID(track) {
			return fmt.Errorf("evaluation method track is invalid")
		}
		if _, duplicate := seenTracks[track]; duplicate {
			return fmt.Errorf("evaluation method tracks must be unique")
		}
		if _, applicable := seenApplicableTracks[track]; !applicable {
			return fmt.Errorf("evaluation method live track must be applicable")
		}
		seenTracks[track] = struct{}{}
	}
	seenMetrics := make(map[string]struct{}, len(method.ProducedMetricIDs))
	for _, metric := range method.ProducedMetricIDs {
		if metric == "" || metric != strings.TrimSpace(metric) {
			return fmt.Errorf("evaluation method metric id is invalid")
		}
		if _, duplicate := seenMetrics[metric]; duplicate {
			return fmt.Errorf("evaluation method metric ids must be unique")
		}
		seenMetrics[metric] = struct{}{}
	}
	seenArtifacts := make(map[string]struct{}, len(method.RequiredArtifactIDs))
	for _, artifact := range method.RequiredArtifactIDs {
		if !validMethodID(artifact) {
			return fmt.Errorf("evaluation method artifact id is invalid")
		}
		if _, duplicate := seenArtifacts[artifact]; duplicate {
			return fmt.Errorf("evaluation method artifact ids must be unique")
		}
		seenArtifacts[artifact] = struct{}{}
	}
	return validateAnalysisPlan(method.AnalysisPlan)
}

type CompoundModelBudgetOutcome struct {
	CaseID    string     `json:"case_id"`
	Action    ActionRef  `json:"action"`
	Budget    int        `json:"budget"`
	Score     float64    `json:"score"`
	SliceRefs []SliceRef `json:"slice_refs"`
}

type SharedDomainCurvePoint struct {
	Action    ActionRef `json:"action"`
	Budget    int       `json:"budget"`
	MeanScore float64   `json:"mean_score"`
	CaseCount int       `json:"case_count"`
}

type CompoundModelBudgetReport struct {
	Method                       EvaluationMethodDefinition `json:"method"`
	AnalysisPlan                 AnalysisPlan               `json:"analysis_plan"`
	ActionRefs                   []ActionRef                `json:"action_refs"`
	SliceRefs                    []SliceRef                 `json:"slice_refs"`
	RawSharedDomainCurve         []SharedDomainCurvePoint   `json:"raw_shared_domain_curve"`
	AUDC                         float64                    `json:"audc"`
	NAUC                         float64                    `json:"nauc"`
	Peak                         float64                    `json:"peak"`
	QNC                          float64                    `json:"qnc"`
	MissingCaseActionBudgetCells int                        `json:"missing_case_action_budget_cells"`
}

func R2CompoundModelBudgetMethod() EvaluationMethodDefinition {
	benchmark, found := researchBenchmarkByAdapter("r2-router")
	if !found {
		panic("research benchmark inventory is missing r2-router")
	}
	return researchBenchmarkMethod(benchmark)
}

// ReduceCompoundModelBudget refuses a ragged cohort.  This preserves both the
// supplied ActionRef identity and a truly shared case×action×budget domain.
func ReduceCompoundModelBudget(method EvaluationMethodDefinition, outcomes []CompoundModelBudgetOutcome) (CompoundModelBudgetReport, error) {
	if method.ID != R2CompoundModelBudgetMethodID {
		return CompoundModelBudgetReport{}, fmt.Errorf("compound reducer requires the R2 compound method")
	}
	if err := ValidateEvaluationMethodDefinition(method); err != nil {
		return CompoundModelBudgetReport{}, err
	}
	if len(outcomes) == 0 {
		return CompoundModelBudgetReport{}, fmt.Errorf("compound reducer requires outcomes")
	}
	cases, actions, budgets, slices := make(map[string]struct{}), make(map[string]struct{}), make(map[int]struct{}), make(map[string]struct{})
	cells := make(map[string]float64)
	expectedSlices := make(map[string]struct{}, len(method.AnalysisPlan.Slices))
	for _, slice := range method.AnalysisPlan.Slices {
		expectedSlices[slice.ID] = struct{}{}
	}
	for _, outcome := range outcomes {
		if !validMethodID(outcome.CaseID) || validateActionRef(outcome.Action) != nil || outcome.Budget <= 0 ||
			math.IsNaN(outcome.Score) || math.IsInf(outcome.Score, 0) || outcome.Score < 0 || outcome.Score > 1 || len(outcome.SliceRefs) == 0 {
			return CompoundModelBudgetReport{}, fmt.Errorf("compound outcome is invalid")
		}
		localSlices := make(map[string]struct{}, len(outcome.SliceRefs))
		for _, slice := range outcome.SliceRefs {
			if err := validateSliceRef(slice); err != nil {
				return CompoundModelBudgetReport{}, err
			}
			if _, duplicate := localSlices[slice.ID]; duplicate {
				return CompoundModelBudgetReport{}, fmt.Errorf("compound outcome slices must be unique")
			}
			localSlices[slice.ID] = struct{}{}
			slices[slice.ID] = struct{}{}
		}
		if len(localSlices) != len(expectedSlices) {
			return CompoundModelBudgetReport{}, fmt.Errorf("compound outcome slices must exactly match the analysis plan")
		}
		for sliceID := range expectedSlices {
			if _, present := localSlices[sliceID]; !present {
				return CompoundModelBudgetReport{}, fmt.Errorf("compound outcome slices must exactly match the analysis plan")
			}
		}
		key := outcome.CaseID + "\x00" + outcome.Action.ID + "\x00" + fmt.Sprint(outcome.Budget)
		if _, duplicate := cells[key]; duplicate {
			return CompoundModelBudgetReport{}, fmt.Errorf("duplicate case×action×budget outcome: %s×%s×%d", outcome.CaseID, outcome.Action.ID, outcome.Budget)
		}
		cells[key] = outcome.Score
		cases[outcome.CaseID], actions[outcome.Action.ID], budgets[outcome.Budget] = struct{}{}, struct{}{}, struct{}{}
	}
	caseIDs, actionIDs, budgetValues, sliceIDs := methodV2SortedStrings(cases), methodV2SortedStrings(actions), methodV2SortedInts(budgets), methodV2SortedStrings(slices)
	if len(cells) != len(caseIDs)*len(actionIDs)*len(budgetValues) {
		return CompoundModelBudgetReport{}, fmt.Errorf("compound model+budget outcomes must form an exact shared case×action×budget domain")
	}
	if len(budgetValues) < 2 {
		return CompoundModelBudgetReport{}, fmt.Errorf("compound AUDC requires at least two shared budget points")
	}
	curve := make([]SharedDomainCurvePoint, 0, len(actionIDs)*len(budgetValues))
	means := make(map[string]float64)
	for _, actionID := range actionIDs {
		for _, budget := range budgetValues {
			total := 0.0
			for _, caseID := range caseIDs {
				total += cells[caseID+"\x00"+actionID+"\x00"+fmt.Sprint(budget)]
			}
			mean := total / float64(len(caseIDs))
			means[actionID+"\x00"+fmt.Sprint(budget)] = mean
			curve = append(curve, SharedDomainCurvePoint{Action: ActionRef{SchemaVersion: EvaluationMethodContractVersion, ID: actionID}, Budget: budget, MeanScore: mean, CaseCount: len(caseIDs)})
		}
	}
	audc, peak, qnc := 0.0, 0.0, 0.0
	for _, actionID := range actionIDs {
		for index := 0; index < len(budgetValues)-1; index++ {
			lower, upper := budgetValues[index], budgetValues[index+1]
			audc += float64(upper-lower) * (means[actionID+"\x00"+fmt.Sprint(lower)] + means[actionID+"\x00"+fmt.Sprint(upper)]) / 2
		}
		terminal := means[actionID+"\x00"+fmt.Sprint(budgetValues[len(budgetValues)-1])]
		qnc += terminal
	}
	for _, mean := range means {
		if mean > peak {
			peak = mean
		}
	}
	nauc := audc / (float64(len(actionIDs)) * float64(budgetValues[len(budgetValues)-1]-budgetValues[0]))
	qnc /= float64(len(actionIDs))
	actionRefs, sliceRefs := make([]ActionRef, len(actionIDs)), make([]SliceRef, len(sliceIDs))
	for index, id := range actionIDs {
		actionRefs[index] = ActionRef{SchemaVersion: EvaluationMethodContractVersion, ID: id}
	}
	for index, id := range sliceIDs {
		sliceRefs[index] = SliceRef{SchemaVersion: EvaluationMethodContractVersion, ID: id}
	}
	return CompoundModelBudgetReport{Method: method, AnalysisPlan: method.AnalysisPlan, ActionRefs: actionRefs, SliceRefs: sliceRefs, RawSharedDomainCurve: curve, AUDC: audc, NAUC: nauc, Peak: peak, QNC: qnc}, nil
}

// ReduceSealedMethodReports is the only publication path for v2 method
// reports.  It deliberately starts from validated raw records, not a worker
// aggregate, so curves and summary metrics cannot be forged independently.
func ReduceSealedMethodReports(methods methodRecordAttestation) ([]CompoundModelBudgetReport, error) {
	reports := make([]CompoundModelBudgetReport, 0, 1)
	if len(methods.R2Outcomes) == 0 {
		return reports, nil
	}
	report, err := ReduceCompoundModelBudget(R2CompoundModelBudgetMethod(), methods.R2Outcomes)
	if err != nil {
		return nil, err
	}
	return append(reports, report), nil
}

func validateSealedMethodReports(actual []CompoundModelBudgetReport, methods methodRecordAttestation) error {
	if actual == nil {
		return fmt.Errorf("method_reports cannot be null")
	}
	expected, err := ReduceSealedMethodReports(methods)
	if err != nil {
		return err
	}
	if !reflect.DeepEqual(actual, expected) {
		return fmt.Errorf("method_reports do not match server-reduced raw method coordinates")
	}
	return nil
}

func methodV2SortedStrings(values map[string]struct{}) []string {
	result := make([]string, 0, len(values))
	for value := range values {
		result = append(result, value)
	}
	sort.Strings(result)
	return result
}

func methodV2SortedInts(values map[int]struct{}) []int {
	result := make([]int, 0, len(values))
	for value := range values {
		result = append(result, value)
	}
	sort.Ints(result)
	return result
}

File: src/vllm-sr/cli/evaluation/method_contract_v2.py
"""Clean-break contracts for executable, gradeable evaluation methods.

This module deliberately does not translate the pre-v2 method records.  A
method is either described and reduced under this contract or it is not a v2
method.  That makes a report's methodological claims independently auditable.
"""

from __future__ import annotations

from collections.abc import Iterable
from itertools import pairwise
from math import isfinite
from typing import Literal

from pydantic import Field, StrictBool, field_validator, model_validator

from cli.evaluation.constants import TRACK_IDS
from cli.evaluation.contract_primitives import StrictModel
from cli.evaluation.contract_validation import validate_portable_id

EVALUATION_METHOD_CONTRACT_VERSION = "evaluation-method.v2"
COMPOUND_MODEL_BUDGET_METHOD_ID = "r2.compound-model-budget.v2"
_MIN_SHARED_BUDGET_POINTS = 2


class ActionRef(StrictModel):
    """A stable action identity; labels must never be used as identities."""

    schema_version: Literal[EVALUATION_METHOD_CONTRACT_VERSION]
    id: str

    _id = field_validator("id")(validate_portable_id)


class SliceRef(StrictModel):
    """A stable cohort label included in the analysis contract and report."""

    schema_version: Literal[EVALUATION_METHOD_CONTRACT_VERSION]
    id: str

    _id = field_validator("id")(validate_portable_id)


class AnalysisPlan(StrictModel):
    """The unit, dependence cluster, slices, curve domain and null policy."""

    schema_version: Literal[EVALUATION_METHOD_CONTRACT_VERSION]
    id: str
    analysis_unit: str
    cluster_unit: str
    slices: tuple[SliceRef, ...]
    curve_domain: Literal["shared_budget", "not_applicable"]
    missingness: Literal["fail_closed"]

    _id = field_validator("id")(validate_portable_id)

    @model_validator(mode="after")
    def unique_slices(self) -> AnalysisPlan:
        ids = tuple(slice_ref.id for slice_ref in self.slices)
        if not ids or len(ids) != len(set(ids)):
            raise ValueError("analysis plan slices must be non-empty and unique")
        return self

    @model_validator(mode="after")
    def non_blank_units(self) -> AnalysisPlan:
        if not self.analysis_unit.strip() or not self.cluster_unit.strip():
            raise ValueError("analysis plan units must be non-blank")
        return self


class EvaluationMethodPlugin(StrictModel):
    """Versioned declaration of what a method can honestly execute and grade."""

    schema_version: Literal[EVALUATION_METHOD_CONTRACT_VERSION]
    id: str
    version: Literal[EVALUATION_METHOD_CONTRACT_VERSION]
    status: Literal[
        "native-qualified", "exploratory-import", "data-required", "blocked"
    ]
    execution_owner: Literal["server", "worker", "provider", "benchmark_native"]
    input_schema: str
    export_schema: str
    live_input_complete: StrictBool
    live_grader: StrictBool
    applicable_tracks: tuple[str, ...]
    live_tracks: tuple[str, ...]
    produced_metric_ids: tuple[str, ...]
    evidence_ceiling: Literal["E0", "E1", "E2", "E3", "E4", "E5"]
    native_parity: Literal["native", "source_qualified", "none"]
    required_artifact_ids: tuple[str, ...]
    analysis_plan: AnalysisPlan

    _id = field_validator("id")(validate_portable_id)

    @field_validator("input_schema", "export_schema")
    @classmethod
    def portable_schema_id(cls, value: str) -> str:
        return validate_portable_id(value)

    @model_validator(mode="after")
    def validate_live_claim(self) -> EvaluationMethodPlugin:
        if self.version != EVALUATION_METHOD_CONTRACT_VERSION:
            raise ValueError("method plugin version must match the v2 contract")
        if (
            not self.applicable_tracks
            or len(self.applicable_tracks) != len(set(self.applicable_tracks))
            or len(self.live_tracks) != len(set(self.live_tracks))
            or not self.produced_metric_ids
            or len(self.produced_metric_ids) != len(set(self.produced_metric_ids))
        ):
            raise ValueError(
                "method metric ids must be non-empty and all identities must be unique"
            )
        if any(
            not metric_id or metric_id.strip() != metric_id
            for metric_id in self.produced_metric_ids
        ):
            raise ValueError("method metric ids must be trimmed")
        known_tracks = frozenset(TRACK_IDS)
        if any(track not in known_tracks for track in self.applicable_tracks):
            raise ValueError("method applicable tracks must be known")
        if any(track not in known_tracks for track in self.live_tracks):
            raise ValueError("method live tracks must be known")
        if not set(self.live_tracks).issubset(self.applicable_tracks):
            raise ValueError("method live tracks must be applicable")
        if not self.required_artifact_ids or len(self.required_artifact_ids) != len(
            set(self.required_artifact_ids)
        ):
            raise ValueError("method required artifacts must be non-empty and unique")
        for artifact_id in self.required_artifact_ids:
            validate_portable_id(artifact_id)
        if self.status == "native-qualified" and (
            not self.live_input_complete or not self.live_grader or not self.live_tracks
        ):
            raise ValueError(
                "native-qualified methods require complete live input, grading, and tracks"
            )
        if self.status != "native-qualified" and (
            self.live_input_complete or self.live_grader
        ):
            raise ValueError(
                "non-qualified methods cannot claim complete live execution"
            )
        if (
            self.native_parity == "native"
            and self.execution_owner != "benchmark_native"
        ):
            raise ValueError("native parity requires benchmark-native execution")
        return self


class CompoundModelBudgetOutcome(StrictModel):
    """One graded action outcome at one point on the shared budget domain."""

    case_id: str
    action: ActionRef
    budget: int = Field(gt=0)
    score: float = Field(ge=0, le=1, allow_inf_nan=False)
    slice_refs: tuple[SliceRef, ...]

    _case_id = field_validator("case_id")(validate_portable_id)

    @model_validator(mode="after")
    def has_unique_slices(self) -> CompoundModelBudgetOutcome:
        ids = tuple(slice_ref.id for slice_ref in self.slice_refs)
        if not ids or len(ids) != len(set(ids)):
            raise ValueError("compound outcomes require unique non-empty slices")
        return self


class SharedDomainCurvePoint(StrictModel):
    action: ActionRef
    budget: int = Field(gt=0)
    mean_score: float = Field(ge=0, le=1, allow_inf_nan=False)
    case_count: int = Field(gt=0)


class CompoundModelBudgetReport(StrictModel):
    """Report payload for R2: identity, analysis plan, curve and missingness."""

    method: EvaluationMethodPlugin
    analysis_plan: AnalysisPlan
    action_refs: tuple[ActionRef, ...]
    slice_refs: tuple[SliceRef, ...]
    raw_shared_domain_curve: tuple[SharedDomainCurvePoint, ...]
    audc: float = Field(ge=0, allow_inf_nan=False)
    nauc: float = Field(ge=0, le=1, allow_inf_nan=False)
    peak: float = Field(ge=0, le=1, allow_inf_nan=False)
    qnc: float = Field(ge=0, le=1, allow_inf_nan=False)
    missing_case_action_budget_cells: int = Field(ge=0)


R2_COMPOUND_MODEL_BUDGET_PLUGIN = EvaluationMethodPlugin(
    schema_version=EVALUATION_METHOD_CONTRACT_VERSION,
    id=COMPOUND_MODEL_BUDGET_METHOD_ID,
    version=EVALUATION_METHOD_CONTRACT_VERSION,
    status="exploratory-import",
    execution_owner="server",
    input_schema="r2-compound-input",
    export_schema="r2-compound-report",
    live_input_complete=False,
    live_grader=False,
    applicable_tracks=("routing", "model_pool", "joint", "capacity"),
    live_tracks=(),
    produced_metric_ids=(
        "r2.compound_model_budget.audc",
        "r2.compound_model_budget.nauc",
        "r2.compound_model_budget.peak",
        "r2.compound_model_budget.qnc",
    ),
    evidence_ceiling="E0",
    native_parity="source_qualified",
    required_artifact_ids=("curves",),
    analysis_plan=AnalysisPlan(
        schema_version=EVALUATION_METHOD_CONTRACT_VERSION,
        id="r2-compound-case-action-budget",
        analysis_unit="case_action_budget",
        cluster_unit="case",
        slices=(
            SliceRef(
                schema_version=EVALUATION_METHOD_CONTRACT_VERSION,
                id="all",
            ),
        ),
        curve_domain="shared_budget",
        missingness="fail_closed",
    ),
)


def _compound_domain(
    rows: tuple[CompoundModelBudgetOutcome, ...],
    method: EvaluationMethodPlugin,
) -> tuple[
    tuple[str, ...],
    tuple[str, ...],
    tuple[int, ...],
    tuple[str, ...],
    dict[tuple[str, str, int], float],
]:
    cases = tuple(sorted({row.case_id for row in rows}))
    actions = tuple(sorted({row.action.id for row in rows}))
    budgets = tuple(sorted({row.budget for row in rows}))
    slices = tuple(
        sorted({slice_ref.id for row in rows for slice_ref in row.slice_refs})
    )
    seen: set[tuple[str, str, int]] = set()
    cells: dict[tuple[str, str, int], float] = {}
    expected_slice_ids = {slice_ref.id for slice_ref in method.analysis_plan.slices}
    for row in rows:
        key = (row.case_id, row.action.id, row.budget)
        if key in seen:
            raise ValueError(
                "duplicate case x action x budget outcome: " + " x ".join(map(str, key))
            )
        seen.add(key)
        cells[key] = row.score
        if {slice_ref.id for slice_ref in row.slice_refs} != expected_slice_ids:
            raise ValueError(
                "compound outcome slices must exactly match the analysis plan"
            )
    expected = len(cases) * len(actions) * len(budgets)
    if len(cells) != expected:
        raise ValueError(
            "compound model+budget outcomes must form an exact shared "
            "case x action x budget domain"
        )
    return cases, actions, budgets, slices, cells


def _compound_curve(
    cases: tuple[str, ...],
    actions: tuple[str, ...],
    budgets: tuple[int, ...],
    cells: dict[tuple[str, str, int], float],
) -> tuple[list[SharedDomainCurvePoint], dict[tuple[str, int], float]]:
    curve: list[SharedDomainCurvePoint] = []
    score_by_action_budget: dict[tuple[str, int], float] = {}
    for action_id in actions:
        for budget in budgets:
            values = [cells[(case_id, action_id, budget)] for case_id in cases]
            mean_score = sum(values) / len(values)
            score_by_action_budget[(action_id, budget)] = mean_score
            curve.append(
                SharedDomainCurvePoint(
                    action=ActionRef(
                        schema_version=EVALUATION_METHOD_CONTRACT_VERSION,
                        id=action_id,
                    ),
                    budget=budget,
                    mean_score=mean_score,
                    case_count=len(cases),
                )
            )
    return curve, score_by_action_budget


def _compound_summary(
    actions: tuple[str, ...],
    budgets: tuple[int, ...],
    score_by_action_budget: dict[tuple[str, int], float],
) -> tuple[float, float, float, float]:
    if len(budgets) < _MIN_SHARED_BUDGET_POINTS:
        raise ValueError("compound AUDC requires at least two shared budget points")
    audc = 0.0
    for action_id in actions:
        for lower, upper in pairwise(budgets):
            audc += (
                (upper - lower)
                * (
                    score_by_action_budget[(action_id, lower)]
                    + score_by_action_budget[(action_id, upper)]
                )
                / 2
            )
    budget_span = budgets[-1] - budgets[0]
    nauc = audc / (len(actions) * budget_span)
    peak = max(score_by_action_budget.values())
    qnc = sum(
        score_by_action_budget[(action_id, budgets[-1])] for action_id in actions
    ) / len(actions)
    if not all(isfinite(value) for value in (audc, nauc, peak, qnc)):
        raise ValueError("compound reduction produced a non-finite metric")
    return audc, nauc, peak, qnc


def reduce_compound_model_budget(
    outcomes: Iterable[CompoundModelBudgetOutcome],
    *,
    method: EvaluationMethodPlugin = R2_COMPOUND_MODEL_BUDGET_PLUGIN,
) -> CompoundModelBudgetReport:
    """Reduce R2 with exact rectangular cardinality and a shared raw domain.

    AUDC is the sum of trapezoids across each action's score/budget curve.
    nAUC divides AUDC by the common budget span and action count. QNC is the
    mean score at the largest common budget; it intentionally never imputes a
    missing arm, case, or budget cell.
    """

    if method.id != COMPOUND_MODEL_BUDGET_METHOD_ID:
        raise ValueError("compound reducer requires the R2 compound method plugin")
    rows = tuple(outcomes)
    if not rows:
        raise ValueError("compound reducer requires outcomes")
    cases, actions, budgets, slices, cells = _compound_domain(rows, method)
    curve, score_by_action_budget = _compound_curve(cases, actions, budgets, cells)
    audc, nauc, peak, qnc = _compound_summary(actions, budgets, score_by_action_budget)
    return CompoundModelBudgetReport(
        method=method,
        analysis_plan=method.analysis_plan,
        action_refs=tuple(
            ActionRef(
                schema_version=EVALUATION_METHOD_CONTRACT_VERSION,
                id=action_id,
            )
            for action_id in actions
        ),
        slice_refs=tuple(
            SliceRef(
                schema_version=EVALUATION_METHOD_CONTRACT_VERSION,
                id=slice_id,
            )
            for slice_id in slices
        ),
        raw_shared_domain_curve=tuple(curve),
        audc=audc,
        nauc=nauc,
        peak=peak,
        qnc=qnc,
        missing_case_action_budget_cells=0,
    )
